iceprog/iceprog.c

```diff
@@ -407,24 +422,124 @@ int main(int argc, char **argv)
 		case 'v':
 			verbose = true;
 			break;
-		default:
+		case -2:
 			help(argv[0]);
+			return EXIT_SUCCESS;
+		default:
+			/* error message has already been printed */
+			fprintf(stderr, "Try `%s --help' "
+					"for more information.\n", argv[0]);
+			return EXIT_FAILURE;
 		}
 	}
 
 	if (read_mode + check_mode + prog_sram + test_mode > 1)
-		help(argv[0]);
+		errx(EXIT_FAILURE,
+		     "options `-r'/`-R', `-c', `-S', and `-t' are mutually "
+		     "exclusive");
 
 	if (bulk_erase && dont_erase)
-		help(argv[0]);
-
-	if (optind+1 != argc && !test_mode) {
-		if (bulk_erase && optind == argc)
-			filename = "/dev/null";
-		else
-			help(argv[0]);
-	} else
+		errx(EXIT_FAILURE,
+		     "options `-b' and `-n' are mutually exclusive");
+
+	if (bulk_erase && (read_mode || check_mode || prog_sram || test_mode))
+		errx(EXIT_FAILURE,
+		     "option `-b' only valid in programming mode");
+	if (dont_erase && (read_mode || check_mode || prog_sram || test_mode))
+		errx(EXIT_FAILURE,
+		     "option `-n' only valid in programming mode");
+
+	if (rw_offset != 0 && prog_sram)
+		errx(EXIT_FAILURE, "option `-o' not supported in SRAM mode");
+	if (rw_offset != 0 && test_mode)
+		errx(EXIT_FAILURE, "option `-o' not supported in test mode");
+
+	if (optind + 1 == argc) {
+		if (test_mode) {
+			warnx("test mode doesn't take a file name");
+			fprintf(stderr, "Try `%s --help' "
+					"for more information.\n", argv[0]);
+			return EXIT_FAILURE;
+		}
 		filename = argv[optind];
+	} else if (optind != argc) {
+		warnx("too many arguments");
+		fprintf(stderr, "Try `%s --help' "
+				"for more information.\n", argv[0]);
+		return EXIT_FAILURE;
+	} else if (bulk_erase) {
+		filename = "/dev/null";
+	} else if (!test_mode) {
+		warnx("missing argument");
+		fprintf(stderr, "Try `%s --help' "
+				"for more information.\n", argv[0]);
+		return EXIT_FAILURE;
+	}
+
+	/* open input/output file in advance
+	   so we can fail before initializing the hardware */
+
+	FILE *f = NULL;
+	long file_size = -1;
+
+	if (test_mode)
```

Comment:
Thanks, I had missed the `-t` option. A dry run mode would be nice, but it is not required. I also just want to capture that as we were discussing this, you thought about documenting that the test mode can be used to clear the reset pin, and that you thought you needed the reverse of this as well.